src/Umbraco.New.Cms.Core/Services/Installer/UpgradeService.cs

```diff
@@ -0,0 +1,48 @@
+﻿using Microsoft.Extensions.Logging;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Services;
+using Umbraco.New.Cms.Core.Installer;
+
+namespace Umbraco.New.Cms.Core.Services.Installer;
+
+public class UpgradeService : IUpgradeService
+{
+    private readonly UpgradeStepCollection _upgradeSteps;
+    private readonly IRuntimeState _runtimeState;
+    private readonly ILogger<UpgradeService> _logger;
+
+    public UpgradeService(
+        UpgradeStepCollection upgradeSteps,
+        IRuntimeState runtimeState,
+        ILogger<UpgradeService> logger)
+    {
+        _upgradeSteps = upgradeSteps;
+        _runtimeState = runtimeState;
+        _logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public async Task Upgrade()
+    {
+        if (_runtimeState.Level != RuntimeLevel.Upgrade)
+        {
+            throw new InvalidOperationException(
+                $"Runtime level must be Upgrade to upgrade but was: {_runtimeState.Level}");
+        }
+
+        foreach (IUpgradeStep step in _upgradeSteps)
```

Comment:
Move steps execution from `Upgrade()` into a new `RunSteps()` method (like `InstallService`) and possibly wrap it in a try/catch (see `InstallService` comments).